check-cloudwatch-logs/lib/check-cloudwatch-logs.go

```diff
@@ -0,0 +1,211 @@
+	CriticalOver  int    `short:"c" long:"critical-over" value-name:"CRITICAL" description:"Trigger a critical if matched lines is over a number"`
+	StateDir      string `short:"s" long:"state-dir" value-name:"DIR" description:"Dir to keep state files under"`
+	ReturnContent bool   `short:"r" long:"return" description:"Output matched lines"`
+}
+
+// Do the plugin
+func Do() {
+	ckr := run(os.Args[1:])
+	ckr.Name = "CloudWatch Logs"
+	ckr.Exit()
+}
+
+type cloudwatchLogsPlugin struct {
+	Service   cloudwatchlogsiface.CloudWatchLogsAPI
+	StateFile string
+	*logOpts
+}
+
+func newCloudwatchLogsPlugin(opts *logOpts, args []string) (*cloudwatchLogsPlugin, error) {
+	var err error
+	p := &cloudwatchLogsPlugin{logOpts: opts}
+	p.Service, err = createService(opts)
+	if err != nil {
+		return nil, err
+	}
+	if p.StateDir == "" {
+		workdir := pluginutil.PluginWorkDir()
+		p.StateDir = filepath.Join(workdir, "check-cloudwatch-logs")
+	}
+	p.StateFile = getStateFile(p.StateDir, opts.LogGroupName, args)
+	return p, nil
+}
+
+var stateRe = regexp.MustCompile(`[^-a-zA-Z0-9_.]`)
+
+func getStateFile(stateDir, logGroupName string, args []string) string {
+	return filepath.Join(
+		stateDir,
+		fmt.Sprintf(
+			"%s-%x.json",
+			strings.TrimLeft(stateRe.ReplaceAllString(logGroupName, "_"), "_"),
+			md5.Sum([]byte(os.Getenv("AWS_PROFILE")+" "+strings.Join(args, " "))),
+		),
+	)
+}
+
+func createService(opts *logOpts) (*cloudwatchlogs.CloudWatchLogs, error) {
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, err
+	}
+	config := aws.NewConfig()
+	if opts.AccessKeyID != "" && opts.SecretAccessKey != "" {
+		config = config.WithCredentials(
+			credentials.NewStaticCredentials(opts.AccessKeyID, opts.SecretAccessKey, ""),
+		)
+	}
+	if opts.Region != "" {
+		config = config.WithRegion(opts.Region)
+	}
+	return cloudwatchlogs.New(sess, config), nil
+}
+
+type logState struct {
+	NextToken *string
+	StartTime *int64
+}
+
+func (p *cloudwatchLogsPlugin) collect() ([]string, error) {
+	var nextToken *string
+	var startTime *int64
+	s, err := p.loadState()
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+	} else {
+		if s.StartTime != nil && *s.StartTime > time.Now().Add(-time.Hour).Unix()*1000 {
+			nextToken = s.NextToken
+			startTime = s.StartTime
+		}
+	}
+	if startTime == nil {
+		startTime = aws.Int64(time.Now().Add(-1*time.Minute).Unix() * 1000)
+	}
+	var messages []string
+	for {
+		output, err := p.Service.FilterLogEvents(&cloudwatchlogs.FilterLogEventsInput{
+			StartTime:     startTime,
+			LogGroupName:  aws.String(p.LogGroupName),
+			NextToken:     nextToken,
+			FilterPattern: aws.String(p.Pattern),
+		})
+		if err != nil {
+			return nil, err
+		}
+		for _, event := range output.Events {
+			messages = append(messages, *event.Message)
+			startTime = aws.Int64(*event.Timestamp + 1)
+		}
+		if output.NextToken != nil {
+			nextToken = output.NextToken
+		}
+		if nextToken != nil {
```

Comment:
That's expected case and is OK because paging finishes when `output.NextToken` is nil. For the next execution the last used NextToken is used and the messages are filtered by startTime.